evalml/data_checks/datetime_format_data_check.py

```diff
@@ -118,21 +179,66 @@ def validate(self, X, y):
             )
             return messages
 
-        if not inferred_freq:
-            col_name = (
-                self.datetime_column
-                if self.datetime_column != "index"
-                else "either index"
-            )
+        col_name = (
+            self.datetime_column if self.datetime_column != "index" else "either index"
+        )
+
+        is_increasing = pd.DatetimeIndex(datetime_values).is_monotonic_increasing
+
+        nan_columns = datetime_values.isna().any()
+        if nan_columns:
             messages.append(
                 DataCheckError(
-                    message=f"No frequency could be detected in {col_name}, possibly due to uneven intervals.",
+                    message=f"Input datetime column ({col_name}) contains NaN values. Please impute NaN values or drop these rows.",
                     data_check_name=self.name,
-                    message_code=DataCheckMessageCode.DATETIME_HAS_UNEVEN_INTERVALS,
+                    message_code=DataCheckMessageCode.DATETIME_HAS_NAN,
                 ).to_dict()
             )
+            # NaN values are a more significant error than missing values and will break the data check logic, so we return here
+            return messages
+
+        if not inferred_freq:
 
-        if not (pd.DatetimeIndex(datetime_values).is_monotonic_increasing):
+            # Check for only one row per datetime
+            duplicate_dates = pd.Series(datetime_values).diff(1) == pd.Timedelta(0)
+            if duplicate_dates.any():
+                messages.append(
+                    DataCheckError(
+                        message=f"{col_name} has more than one row with the same datetime value.",
+                        data_check_name=self.name,
+                        message_code=DataCheckMessageCode.DATETIME_HAS_REDUNDANT_ROW,
+                    ).to_dict()
+                )
+                # drop any duplicates before continuing
+                datetime_values = datetime_values[~duplicate_dates]
+
+            frequencies = []
+            for i in range(len(datetime_values) - 2):
+                freq = pd.infer_freq(datetime_values[i : i + 3])
```

Comment:
hypernit: I think 3 should be assigned to some variable `interval_size` or something with the previous line referencing that `interval_size`, e.g. `- (interval_size-1)`

Curious, though, why not go with the `num_disruptions = len(set(datetime_values.diff(1)))`?  I'll have to check your testing, but I'm not sure how "fuzzy" the frequency inference is, either, like if the given datetimes are separated by basically a day but with a few seconds difference.

Reply:
I like these suggestions! This method of checking for disruptions stems from before I knew about the `diff` function, so I'll update this if it work.

Reply:
After much experimentation, it seems like `num_disruptions = len(set(datetime_values.diff(1)))` changes the distinction between missing vs uneven errors in a way we probably don't want. I'm going to add the `interval_size`, but leave out the other change, at least for now.